src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+the same way declaring a final instance variable of the underlying type
+would in a wrapper class.
+It's in scope in the extension type body, and
+you can access it using its name as a getter:
+
+- Within the extension type body: `this.i`.
+- Outside with a property extraction: `e.i` (where `e` is the object of an extension type). 
+
+Extension types declarations can also include [type parameters](generics)
+just like classes:
+
+```dart
+extension type E<T>(int i) {
+  // ...
+}
+```
+
+### Constructors
+
+You can optionally declare constructors in an extension type's body.
+A constructor must be named, because the the representation declaration
+already acts as an unnamed constructor for the extension type
+that initializes the represenation object.
+Any additional constructors must declare the representation object's
+instance variable in its initializer list:
+
+```dart
+// 'E' has a named constructor 'n',
+// as well as its built-in unnamed constructor
+extension type E(int i) {
+  E.n(this.i, String foo);
+}
+
+void main() {
+  var named = E.n(3, "Hello");
+  var unnamed = E(4);
+}
+```
+
+### Members
+
+Declare members in the body of an extension type to define its interface. 
+Extension type member declarations are identical to class members.
+Members can be methods, getters, setters, or operators:
+
+```dart
+extension type NumberE(int value) {
+  // Operator:
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+  // Getter:
+  NumberE get i => value;
+  // Method:
+  bool isValid() => !value.isNegative;
+}
+```
+
+Members of the representation type are not available to the extension type
+by default. If you want a member from the representation type to be available
+to the extension type, you must write a declaration for it
+in the extension type definition, like the `operator +` in `NumberE`.
+You also can define new members unrelated to the representation type,
+like the `i` getter and `isValid` method.
+
+Extension types can't declare instance fields (unless [`external`][]),
+or [abstract members][].
+
+### Implements
+
+You can optionally use the `implements` clause to introduce
+a subtype relationship on an extension type, making *all* members
+of the representation type available to the extension type:
+
+```dart
+extension type NumberI(int i) 
+  implements int{
+  // ...
+}
+```
+
+An extension type can only implement its representation type
+or a supertype of its representation type.
+
+This is technically NOT an inheritance relationship
+like the one `implements` introduces for classes,
+but rather an [applicability][] relationship like that between an
+extension method and its `on` type.
+Members that apply to the supertype are applicable to the subtype,
+unless the supertype has a more specific definition for the member.
+
+#### `@redeclare`
+
+Replacing a method that shares a name with a member of the supertype
+is not an override relationship for extension types, but a *redeclaration*.
```

Comment:
I like the explanation but yes probably too much to say here :/ (I still added some of it though)

Where are annotations typically defined? It would be great to have this explanation wherever that it, and point to it from here.